### PR TITLE
Added pretty-print function

### PR DESCRIPTION
**Pull request by gkiar**
This takes a descriptor and makes nice help text. The motivation was to use this for updating AFNI help text, and we'd eventually build an importer from this help-text format. I still need to finish all the input parameter information, but see the example below.

```
$ bosh pprint boutiques/schema/examples/test_pretty_print.json
================================================================================

Tool name: Example Boutiques Tool (ver: 0.0.1)
Tool description: This property describes the tool or application
Tags: domain, foo, programming-language

Command-line:
  exampleTool1.py [CONFIG_FILE] [STRING_INPUT_LIST] [STRING_INPUT]
                  [FLAG_INPUT] [NUMBER_INPUT] [ENUM_INPUT]
                  [FILE_INPUT] [LOG] [INT_LIST_INPUT]

================================================================================

Container Information:
	Image: boutiques/example1:test
	Type: docker

================================================================================

Output Files:
	Name: Log file (Required)
	Format: log-[CONFIG_NUMBER]-[STRING_INPUT].txt
	Filename depends on Input IDs: str_input, config_num
	Stripped extensions (before substitution): .txt, .csv

	Name: outfiles (Optional)
	Format: output/*_exampleOutputTag.resultType

	Name: Configuration file (Required)
	Format: ./config.txt
	Template:
	 # This is a demo configuration file
	 numInput=[CONFIG_NUMBER]
	 strInput=[STRING_INPUT]

================================================================================

Input Groups:
	Name: Example Group 1
	Type: Mutually Exclusive, One is Required
	Group Member IDs: num_input, enum_input

================================================================================

Error Codes:
	Return Code: 2
	Description: File does not exist!

================================================================================

positional arguments:
  {4,5,8}               ID: config_num
                        Value Key: [CONFIG_NUMBER]
                        Type: Number
                        List: False
                        Optional: False
                        Integer: False
                        Range: [N/A, N/A]
                        Default Value: 4
  [FILE_INPUT]          ID: file_input
                        Value Key: [FILE_INPUT]
                        Type: File
                        List: False
                        Optional: True
  [LOG]                 ID: logfile
                        Value Key: [LOG]
                        Type: None
                        List: False
                        Optional: False
                        Description: The output log file from the example tool

optional arguments:
  -i [STRING_INPUT_LIST]
                        ID: str_input_list
                        Value Key: [STRING_INPUT_LIST]
                        Type: String
                        List: True
                        Optional: False
                        List Length: 1 - 3
                        Default Value: aStringValue
                        Description: Describe the use and meaning of the parameter
  -s [STRING_INPUT]     ID: str_input
                        Value Key: [STRING_INPUT]
                        Type: String
                        List: False
                        Optional: False
                        Default Value: aStringValue
                        Description: Describe the use and meaning of the parameter
  -l [INT_LIST_INPUT]   ID: list_int_input
                        Value Key: [INT_LIST_INPUT]
                        Type: Number
                        List: True
                        Optional: False
                        List Length: >= 1
                        Integer: True
                        Range: [N/A, N/A]
                        Description: Describe the use and meaning of the parameter
  -n [NUMBER_INPUT]     ID: num_input
                        Value Key: [NUMBER_INPUT]
                        Type: Number
                        List: False
                        Optional: True
                        Integer: False
                        Range: (N/A, 1]
  -e {val1,val2,val3}   ID: enum_input
                        Value Key: [ENUM_INPUT]
                        Type: String
                        List: False
                        Optional: True
                        Value Dependency: 
                         Value    Disables    Requires
                        -------  ----------  ----------
                        val1     num_input   file_input
                        val2
                        val3
  -f [FLAG_INPUT]       ID: flag_input
                        Value Key: [FLAG_INPUT]
                        Type: Flag
                        List: False
                        Optional: True
                        Disables: num_input
                        Requires: file_input
  -c [CONFIG_FILE]      ID: config_file
                        Value Key: [CONFIG_FILE]
                        Type: None
                        List: False
                        Optional: False

```

### CODE REVIEW

File: tools/python/boutiques/bosh.py [hunk 1]
```diff
@@ -20,6 +20,19 @@
...
 
 
+def prettyprint(*params):
+    parser = ArgumentParser("Boutiques pretty-print for generating help text")
+    parser.add_argument("descriptor", action="store",
+                        help="The Boutiques descriptor.")
+    results = parser.parse_args(params)
+
+    from boutiques.prettyprint import pprint
+    with open(results.descriptor, 'r') as fhandle:
```
Comment (erinb90) [inline tools/python/boutiques/bosh.py]: To make pretty-print compatible with JSON objects and Zenodo files, you can do something like:
```
desc = loadJson(results.descriptor)
helpstring = pprint(desc)
```

Comment (gkiar) [reply tools/python/boutiques/bosh.py]: Great catch - thanks!

**Comment by gkiar** [comment]
Thanks for the feedback, @erinb90 ! My only counter point is I generally like to make classes when objects and modules are either recyclable or multi-functional. In this case, the function is long, but it's kind of just a 1-shot execution where each piece of information is used only once to record it in a more printable format. Would you still recommend breaking this into a class? Happy to heed your advice if so! 😄

**Comment by erinb90** [comment]
It doesn't have to be a class, but you could still break `pprint` into multiple functions to make things more modular. The flow could be something like this:
```
1. create_description(descriptor) ==> returns (sep, name, description, tags)
2. create_lookup_tables(descriptor) ==> returns (clkeys_lut1, clkeys_lut2)
3. parse_cline(descriptor, clkeys_lut2) ==> returns cline
4. init tool_description
5. get_container_info(descriptor) ==> returns container_info
6. if container_info is not None:
        tool_description += ...
# repeat 5,6 for output_info, group_info, res_info and ecod_info
7. create ArgumentParser
   inputs = ...
8. for each clkey in clkeys_lut2:
        add_input_args(parser, clkey, inputs) ==> adds input args to parser and returns parser 
9. helptext = ...
   prettystring = ...
   return prettystring
```

**Comment by gkiar** [comment]
Okeydoke, well it would've been clunky to take part of your advice without putting it in a class, so I've refactored it as suggested, @erinb90 !

**Comment by glatard** [comment]
Really a cool command! I managed to make it crash though, when tools don't have output files:
```
$ bosh pprint zenodo.1895219
Traceback (most recent call last):
  File "/home/glatard/virtualenvs/boutiques/bin/bosh", line 11, in <module>
    sys.exit(bosh())
  File "/home/glatard/virtualenvs/boutiques/lib/python3.6/site-packages/boutiques/bosh.py", line 547, in bosh
    out = prettyprint(*params)
  File "/home/glatard/virtualenvs/boutiques/lib/python3.6/site-packages/boutiques/bosh.py", line 31, in prettyprint
    prettyclass = PrettyPrinter(desc)
  File "/home/glatard/virtualenvs/boutiques/lib/python3.6/site-packages/boutiques/prettyprint.py", line 14, in __init__
    self.createHelpText()
  File "/home/glatard/virtualenvs/boutiques/lib/python3.6/site-packages/boutiques/prettyprint.py", line 17, in createHelpText
    self.createLUT()
  File "/home/glatard/virtualenvs/boutiques/lib/python3.6/site-packages/boutiques/prettyprint.py", line 49, in createLUT
    self.desc.get("output-files"))}
TypeError: can only concatenate list (not "NoneType") to list
```
Once this is fixed I'm happy to merge!

**Comment by gkiar** [comment]
Ah! Thanks, I fixed this robustly, with lines 47:50

**Pull request merged**